### ISSUE TITLE
Line 154-168 in model.py which calculates the last attention-weighted representation r_L and output h_N seems not right.

### ISSUE THREAD

**Issue by jeffacode** [JerrikEph/word_by_word_attn#1]
Here's the code,
```python
def w2w_attn(Premise_out, Hypo_out, seqLen_Premise, seqLen_Hypo, scope=None):
    with tf.variable_scope(scope or 'Attn_layer'):
        attn_cell = AttnCell(self.config.hidden_size, Premise_out, seqLen_Premise)
        attn_cell = rnn_cell.DropoutWrapper(attn_cell, input_keep_prob=self.config.dropout,
                                            output_keep_prob=self.config.dropout)

        _, r_state = tf.nn.dynamic_rnn(attn_cell, Hypo_out, seqLen_Hypo,
                          dtype=Hypo_out.dtype, swap_memory=True)
    return r_state

r_L = w2w_attn(Premise_out, Hypo_out, seqLen_x1, seqLen_x2, scope='w2w_attention')

h_star = tf.tanh(linear([r_L, Hypo_state[1]],       # shape (b_sz, h_sz)
                         self.config.hidden_size, bias=False,
                         scope='linear_trans'))
```
The problem is hypothesis sentence might also be padded with zero vectors, so the real last r_L is not necessarily the one attenLSTM outputs at last time step, as well as Hypo_state[1] which indicates the last h_N here. I believe the right way to do this is first add this [method,](https://danijar.com/variable-sequence-lengths-in-tensorflow/)
```python
def _last_relevant(self, output, length):
    batch_size, max_length, output_size = output.get_shape().as_list()
    index = tf.range(0, batch_size) * max_length + (length - 1)
    flat = tf.reshape(output, [-1, output_size])
    relevant = tf.gather(flat, index)  # (batch_size, output_size)
    return relevant
```
and then, revise line 154-168 to
```python
with tf.variable_scope('w2w_attention'):
    r_state, _ = tf.nn.dynamic_rnn(
        cell=AttnCell(self.config.hidden_dim, Premise_out, Premise_len),
        inputs=Hypo_out,
        sequence_length=Hypo_len,
        dtype=Hypo_out.dtype)

r_L = self._last_relevant(r_state, Hypo_len)

h_N = self._last_relevant(Hypo_out, Hypo_len)

h_star = tf.tanh(linear([r_L, h_N],       # shape (b_sz, h_sz)
                         self.config.hidden_size, bias=False,
                         scope='linear_trans'))
```
Hope it helps:)

**Comment by jingjing-gong** [comment]
You should read the tf.nn.dynamic_rnn code, it have resolved the problem you mentioned. this operation copys the state though if time step exceed the length specified in seqLen_**, so the state returned from dynamic_rnn is valid as if there is no padding.

**Comment by jeffacode** [comment]
I checked the code, and you're right! thanks for your notice:)
